src/ray/stats/metric_exporter_client.cc

```diff
@@ -45,42 +46,133 @@ void MetricExporterDecorator::ReportMetrics(const std::vector<MetricPoint> &poin
+      metric_timeseries_proto->mutable_start_timestamp()->set_seconds(start_time);
+
+      for (const auto &value : tag_values) {
+        metric_timeseries_proto->add_label_values()->set_value(value);
+      };
+
+      auto point_proto = metric_timeseries_proto->add_points();
+      point_proto->mutable_timestamp()->set_seconds(end_time);
+      return point_proto;
+    };
+
+    // Write the `TimeSeries` for the given aggregated data type.
+    switch (view_data.type()) {
+    case opencensus::stats::ViewData::Type::kDouble:
+      for (const auto &row : view_data.double_data()) {
+        std::vector<std::string> tag_values = row.first;
+        double point_value = row.second;
+
+        auto point_proto = make_new_data_point_proto(tag_values);
+        point_proto->set_double_value(point_value);
+      }
+      break;
+    case opencensus::stats::ViewData::Type::kInt64:
+      for (const auto &row : view_data.int_data()) {
+        auto tag_values = row.first;
+        int64_t point_value = row.second;
+
+        auto point_proto = make_new_data_point_proto(tag_values);
+        point_proto->set_int64_value(point_value);
+      }
+      break;
+    case opencensus::stats::ViewData::Type::kDistribution:
+      for (const auto &row : view_data.distribution_data()) {
+        auto tag_values = row.first;
+        opencensus::stats::Distribution dist_value = row.second;
+
+        auto point_proto = make_new_data_point_proto(tag_values);
+
+        // Copy in memory data into `DistributionValue` protobuf.
+        auto distribution_proto = point_proto->mutable_distribution_value();
+        distribution_proto->set_count(dist_value.count());
+        distribution_proto->set_sum(dist_value.count() * dist_value.mean());
+        distribution_proto->set_sum_of_squared_deviation(
+            dist_value.sum_of_squared_deviation());
+
+        // Write the `BucketOption` and `Bucket` data.
+        auto bucket_opt_proto =
+            distribution_proto->mutable_bucket_options()->mutable_explicit_();
+        for (const auto &bound : dist_value.bucket_boundaries().lower_boundaries()) {
+          bucket_opt_proto->add_bounds(bound);
+        }
+        for (const auto &count : dist_value.bucket_counts()) {
+          distribution_proto->add_buckets()->set_count(count);
+        }
+      }
+      break;
+    default:
+      RAY_LOG(FATAL) << "Unknown view data type.";
+      break;
     }
-    // If description and units information is requested from
-    // the metrics agent, append the information.
-    // TODO(sang): It can be inefficient if there are lots of new registered metrics.
-    // We should make it more efficient if there's compelling use cases.
-    if (should_update_description_) {
-      metric_point->set_description(point.measure_descriptor.description());
-      metric_point->set_units(point.measure_descriptor.units());
+
+    // Increment the batch counter, and add another request if necessary.
+    batch_counter++;
```

Comment:
There is no difference with sending them in one batch if we collect all metric point in one loop and send them after that. It's suggested to make full use of rpc async request by sending them in multiple pipelines(collecting in client thread and reporting in rpc thread).

Reply:
- I believe this batch size limit is used for optimizing gRPC message size. gRPC has message limit and recommend to send small messages. 
- The original implementation send requests in just one thread as well, except it's interleaving with the `OpenCensusData -> MetricsPoints` computation.

Reply:
> * I believe this batch size limit is used for optimizing gRPC message size. gRPC has message limit and recommend to send small messages.
> * The original implementation send requests in just one thread as well, except it's interleaving with the `OpenCensusData -> MetricsPoints` computation.

I mean io service/thread pool can be used for rpc client.
```cpp
for () {
   // do something like collecting view data
   if (collected_metric_size >= batch_size) {
      // send them to agent.
  }
}
```
We would not hold all metric proto in memory and it's asynchronous if we do same processing like above.

Reply:
Hmm okay. I can change `ReportMetrics(MetricPoints)` back to the original batching behavior. But leave `ReportMetrics(OpenCensusData)` un-batched because the data doesn't need to be processed. How about this approach?

Reply:
Reached out via Slack. I'm happy to chat more in real-time to resolve the differences.